src/library/scala/collection/immutable/LazyList.scala

```diff
@@ -610,22 +632,26 @@ object LazyList extends LazyListFactory[LazyList] {
     override def tail: LazyList[Nothing] = throw new UnsupportedOperationException("tail of empty lazy list")
     def force: this.type = this
     override def knownSize: Int = 0
-    protected def tailDefined: Boolean = false
-    protected def headDefined: Boolean = false
+    protected[immutable] def tailDefined: Boolean = false
+    protected[immutable] def headDefined: Boolean = false
   }
 
   @SerialVersionUID(3L)
-  final class Cons[A](hd: => A, tl: => LazyList[A]) extends LazyList[A] {
+  final class Cons[A](var hd: () => A, var tl: () => LazyList[A]) extends LazyList[A] {
     private[this] var hdEvaluated: Boolean = false
     private[this] var tlEvaluated: Boolean = false
     override def isEmpty: Boolean = false
     override lazy val head: A = {
       hdEvaluated = true
-      hd
+      val res = hd()
+      hd = null
+      res
```

Comment:
The goal is to let the thunk be garbage collected?

Reply:
Yes.